Adding platform telemetry

walrusmcd opened the pull request:
This PR adds a framework and an implementation for platform telemetry.

The basic idea is to add the concept of platform nuetral telemetry (a pure virtual class) and add the first PAL implementation of a windows class.

more platforms will come later.  

This PR also adds docs and a privacy statement on how telemetry will work.   namely:
- clone and build from source will never tag telemetry events such that windows will listen to them
- official builds will tag telemetry events such that windows *can* listen to them. this will follow the standard CEIP program and optin/optout instructions from the user and/or enterprise.
- official build telemetry will be tagged ON by default.
- we added a global on/off so that an app can choose to turn off all telemetry events for their app, regardless of if the desktop machine is enrolled in windows platform CEIP.

pranavsharma commented:
Some preliminary comments - 
1. Please follow ORT style guide. No camel case naming :-)
2. I guess the session option to turn this off is in progress? 
3. I didn't see a place where this session option would be used. The macro that calls LogRuntimeError would've no knowledge about this option, right?

walrusmcd commented:
1. awesome, i think i have all the style guide fixes, no more camel casing :)
2. i added the enable/disable.  since some of these events span session state, i set it at the global Env .   an app dev can enable and disable them globally, not per session.
3. added the setting into the telemetry for enable/disable.  with the toggle.  and set it to false for the first PR .   we will finish the end to end testing (all the way to the cloud) and then enable them by default.  we also have to configure the build pipeline in order to run things on in the official builds (since clone and build from source will never send windows telemetry )

---
In reply to: [541264286](https://github.com/microsoft/onnxruntime/pull/2109#issuecomment-541264286) [](ancestors = 541264286)

pranavsharma commented on README.md:
```diff
@@ -204,3 +206,8 @@ or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any addi
 # License
 [MIT License](LICENSE)
+
+# Data/Telemetry
```
We must call out that it is disabled by default.

walrusmcd replied:
My idea was to have the DOCS clear that telemetry will be on by default.  then make the code obey that as and when we are ready .   What I want to avoid is have the docs say that they are off by default, and then when we later turn the code on it's not super clear that you need to go an read the docs again (aka breaking change).   in this manner, the docs and "design" are all such that it is clearly messaged telemetry is ON by default.    it's a "bug" that we don't have the code behaving per spec yet.  In that fashion we can "fix the code" later and all be good.   It isnt a breaking change .   you like ?

---
In reply to: [335297148](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335297148) [](ancestors = 335297148)

pranavsharma replied:
I thought we decided to leave the session option for this to be disabled by default in 1.0 release. The session option config should match the docs, right? Or I'm missing something?

faxu replied:
The docs should reflect the current state. To start, it will be OFF by default. When it's out of preview and ready to fully turn on, we will update the documentation to reflect that state. 

For the specific documentation, I suggest:
- main readme page mentions that telemetry collection is implemented in this project and to refer to the detailed data collection page. We don't need to change this documentation when default state changes. 
- data collection page includes privacy policy as well as the documentation on how to enable/disable
- C API page just links to data collection page. We can still have a section heading, but it will contain nothing but the link to the data collection page.

walrusmcd replied:
change the docs to denote it is OFF by default during BETA, and to expect it to ON by default once it graduates from BETA (for transparency)

---
In reply to: [335777476](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335777476) [](ancestors = 335777476)

walrusmcd replied:
I made this change  like you suggested, just a heads up that now we have C API specific docs embedded in the privacy statement.    If we ever have projections on these (like C# or WinRT) we might have to figure out where to put those docs.  I had the turn/on and off in the C_API docs since it was specific to the C_API.

Let me know what you think with the new doc layout !

---
In reply to: [335829030](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335829030) [](ancestors = 335829030)

pranavsharma commented on docs/C_API.md:
```diff
@@ -29,3 +29,12 @@
+# Telemetry
+This project collects usage data and sends it to Microsoft to help improve our products and services. Note however that no data collection is performed by default when using your private builds
+
+Telemetry is turned on by default when using the Official Builds.  This is implemented via 'Platform Telemetry' per vendor platform providers (see telemetry.h).
```
Turned "off" by default. Also, is it better to write the documentation in one place and link to it from everywhere else?

walrusmcd replied:
check out my comments on the other doc on why i like have the "spec" be ON by default, and the code will catch up later when we are ready (bug level fixes) .    as to the docs spread around a bit, i went back and forth on it :) .   pros/cons.

1. other OSS components have multiple docs.  having things like a "privacy statment" are nice.
2. i like the main README being more of a TOC and link off.  that way the root/readme doesn't just bloat and bloat with content.
3. i like the idea of drill down for details.  the top level docs have basic ideas.  and you get to go down deeper (like the c_api doc) to learn how to use the actual C api in this fashion.

what ya think?

---
In reply to: [335297643](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335297643) [](ancestors = 335297643)

pranavsharma replied:
It's ok as long as we don't end up with inconsistent documentation. @faxu can probably review this and give some comments.

pranavsharma commented on include/onnxruntime/core/common/common.h:
```diff
@@ -74,6 +74,9 @@ using common::Status;
 
 std::vector<std::string> GetStackTrace();
+// these is a helper function that gets defined by platform/Telemetry
+void LogRuntimeError(uint32_t sessionId, const common::Status& status, const char* file,
```
session_id

pranavsharma replied:
Should this method be const too?

walrusmcd replied:
it's a not a member function, this one is a global helper.   we only get to use "const" on member functions to denote they can't change member vars. right?

---
In reply to: [335302396](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335302396) [](ancestors = 335302396)

pranavsharma replied:
ah, didn't realize that it wasn't a member function.

pranavsharma commented on onnxruntime/core/platform/posix/env.cc:
```diff
@@ -266,8 +266,14 @@ class PosixEnv : public Env {
 
+  // \brief returns a provider that will handle telemetry on the current platform
+  const Telemetry& GetTelemetryProvider() const override {
+    return telemetryProvider_;
```
telemetry_provider_

pranavsharma commented on onnxruntime/core/platform/windows/env.cc:
```diff
@@ -211,6 +212,11 @@ class WindowsEnv : public Env {
 
+  // \brief returns a provider that will handle telemetry on the current platform
+  const Telemetry& GetTelemetryProvider() const override {
+    return telemetryProvider_;
```
telemetry_provider_

pranavsharma commented on onnxruntime/core/platform/telemetry.h:
```diff
@@ -0,0 +1,55 @@
+                                  const std::unordered_map<std::string, std::string>& model_metadata,
+                                  const std::string& loadedFrom, const std::vector<std::string>& execution_provider_ids) const;
+
+  virtual void LogRuntimeError(uint32_t sessionId, const common::Status& status, const char* file,
```
session_id

pranavsharma commented on onnxruntime/core/session/inference_session.cc:
```diff
@@ -95,6 +95,8 @@ inline std::basic_string<T> GetCurrentTimeString() {
 
 }  // namespace
 
+std::atomic<uint32_t> InferenceSession::global_session_id_{1};
```
Could 2 different users report the same session id in the events? How do we differentiate between them?

walrusmcd replied:
yep.  by 2 users we could have:

- 2 apps on the same user desktop machine login instance
- 2 apps running on different machines

 the telemetry events have an "app session" global that is kept track of by the telemetry engines.    it lets us know which machine/app combo is reporting in.     so session_id is really the "ort session" differentiator inside a specific app instance (aka process ID).

---
In reply to: [335301237](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335301237) [](ancestors = 335301237)

pranavsharma replied:
Is it possible to have 2 different sessions within the same app instance? If yes, can we still differentiate?

walrusmcd replied:
In theory a long lived process could have global_session_id_ wrap around such that every 4B sessions would restart the numbering, and would maintain the same session id.    We actually have a state machine that lives in the cloud @ (xiangting) is helping us to lock down.  This is how we keep track of state and we can use to answer if the wrap around is ok, and also whether or not we need/want a "SessionEnd" event.  I'll file an open issue for this.

---
In reply to: [335779060](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335779060) [](ancestors = 335779060)

walrusmcd replied:
Just to make sure I fully answers the quesiton :) is that what you meant?    Wrap around to 4 billion?  people can totally have 2 OrtSession's in the same app instance, and then we would differentiate using that uint32, right?  (assuming we are ok with wrap around).   another easy fix here is to just make it int64.

---
In reply to: [336570778](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r336570778) [](ancestors = 336570778,335779060)

pranavsharma commented on onnxruntime/core/session/inference_session.h:
```diff
@@ -434,5 +434,13 @@ class InferenceSession {
+  uint32_t total_runs_since_last_;                  // the total number of Run() calls since the last report
+  long long total_run_duration_since_last_;         // the total duration (us) of Run() calls since the last report
+  TimePoint time_sent_last_;                        // the TimePoint of the last report
+  const long long kDurationBetweenSending = 1000* 1000 * 60 * 10;  // duration in (us).  send a report every 10 mins
```
Should this be configurable?

walrusmcd replied:
we could add a feature to make it configurable, but what we have found is that we never really change it.   we've gone a couple of years on some of our events without changing the duration sent.  we don't really want app developers to experiment too much with this number, as it becomes complex as to what guidance we would give them on how to experiment.   it's not a runtime perf thing.   there is some factor of network traffic and queue management, which is why we usually pick numbers like "10 minutes" .  something not too fast as to worry about network congestion or queues, and fast enough that we get regular heartbeat signal.

---
In reply to: [335301525](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335301525) [](ancestors = 335301525)

pranavsharma commented on include/onnxruntime/core/common/common.h:
```diff
@@ -137,16 +140,25 @@ std::vector<std::string> GetStackTrace();
+#define ORT_RETURN_IF_ERROR_SESSIONID_(expr) ORT_RETURN_IF_ERROR_SESSIONID(expr, session_id_)
+#define ORT_RETURN_IF_ERROR(expr) ORT_RETURN_IF_ERROR_SESSIONID(expr, 0)
+
 #define ORT_THROW_IF_ERROR(expr)               \
```
Do we plan to send events for cases when we just do ORT_THROW?

walrusmcd replied:
good catch.  I first coded ORT_RETURN_IF and ORT_THROW_IF .  it covers most all code paths, but not all.   how about i file a tracking item in a second PR to catch all of the macros?

---
In reply to: [335302117](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335302117) [](ancestors = 335302117)

pranavsharma replied:
Sure, that's fine.

walrusmcd replied:
added as a work item

---
In reply to: [335777787](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335777787) [](ancestors = 335777787)

pranavsharma commented on onnxruntime/core/platform/windows/telemetry.h:
```diff
@@ -0,0 +1,57 @@
+
+  void LogProcessInfo() const override;
+
+  void LogSessionCreation(uint32_t session_id, int64_t ir_version, const std::string& model_producer_name,
```
Would it be better to use a struct to pass so many parameters to the function? This way if we add one more thing to the event only the struct needs to change, not the function declarations and definitions across several files.

walrusmcd replied:
i like it.  want me to do that in this PR, or would as second change be ok ?  another thing to keep in mind is that changing this is a big deal.   it's a "schema change" .  so this is not somethin we do all the time, or need to necessarily optimize for.     in our experience we change it *maybe* once a year.   as an example winML is coming on 4 releases now (2 years) and we haven't change the schema yet.

---
In reply to: [335472938](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335472938) [](ancestors = 335472938)

pranavsharma replied:
I think it's going to take us some iterations before we stabilize and it would be nice to avoid the pain of changing across multiple files in the meanwhile.

walrusmcd replied:
i added it as a work item. thx

---
In reply to: [335778701](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335778701) [](ancestors = 335778701)

xxtsxxts commented on onnxruntime/core/platform/windows/telemetry.cc:
```diff
@@ -0,0 +1,192 @@
+                    TraceLoggingString(model_metadata_string.c_str(), "modelMetaData"),
+                    TraceLoggingString(loadedFrom.c_str(), "loadedFrom"),
+                    TraceLoggingString(execution_provider_string.c_str(), "executionProviderIds"));
+}
```
How are we going to get the adapter Luid? So we can get GPU information.
We used to log it in SessionCreation telemetry.

walrusmcd replied:
For now we wont have it.    We have options :
1- add an event to the EP's to output what device was selected.
2- do nothing, this event logs which EP was chosen - so we know when GPU was chosen.  and we can use existing telemetry to know which GPU was on the machine.   it will have a gap if the machine has 2 GPU's (what percentage are those?) 

---
In reply to: [336233302](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r336233302) [](ancestors = 336233302)

xxtsxxts replied:
For option 2, are we going to use the device id to find corresponding GPU?
I'm fine with adding option 1 later, since the GPU usage is not high.

walrusmcd commented on README.md:
```diff
@@ -174,3 +176,8 @@ or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any addi
 # License
 [MIT License](LICENSE)
+
+# Data/Telemetry
```
moved it above license.

---
In reply to: [335830799](https://github.com/microsoft/onnxruntime/pull/2109#discussion_r335830799) [](ancestors = 335830799)

snnn commented:
## Potential Bug in `global_register_count_` Reference Counting

I noticed a potential issue with the `global_register_count_` variable in the `WindowsTelemetry` class introduced in this PR.

### Current Implementation:
```cpp
WindowsTelemetry::WindowsTelemetry() {
  std::lock_guard<OrtMutex> lock(mutex_);
  if (global_register_count_ == 0) {
    HRESULT hr = TraceLoggingRegister(telemetry_provider_handle);
    if (SUCCEEDED(hr)) {
      global_register_count_ += 1;
    }
  }
}

WindowsTelemetry::~WindowsTelemetry() {
  std::lock_guard<OrtMutex> lock(mutex_);
  if (global_register_count_ > 0) {
    global_register_count_ -= 1;
    if (global_register_count_ == 0) {
      TraceLoggingUnregister(telemetry_provider_handle);
    }
  }
}
```

### The Problem:
The counter only increments when it's 0, but decrements on every destructor call. This means:

1. **First object created**: `count = 0 → 1`, registers provider ✅
2. **Second object created**: `count = 1`, does nothing (should be 2) ❌
4. **Second destructor**: `count = 0`, does nothing

The provider gets unregistered while live `WindowsTelemetry` objects still exist, which could cause crashes or undefined behavior.

### Suggested Fix:
```cpp
WindowsTelemetry::WindowsTelemetry() {
  std::lock_guard<OrtMutex> lock(mutex_);
  global_register_count_ += 1;  // Always increment
  if (global_register_count_ == 1) {
    TraceLoggingRegister(telemetry_provider_handle);
  }
}
// Destructor remains the same
```

This same pattern exists in the current codebase in:
- `onnxruntime/core/platform/windows/telemetry.cc` (with `TraceLoggingRegisterEx`)
- `onnxruntime/core/providers/qnn/qnn_telemetry.cc`

This might explain issues users are experiencing with TraceLogging registration/unregistration.
